public/apps/configuration/panels/role-edit.tsx

```diff
@@ -47,24 +58,191 @@ const TITLE_TEXT_DICT = {
   create: 'Create Role',
   edit: 'Edit Role',
   duplicate: 'Duplicate Role',
-}
+};
 
 type OptionSeletion = EuiComboBoxOptionOption[];
 
-function buildPermissionOptions(optionsList: string[]) {
-  return optionsList.map(e => ({ label: e }));
+interface RoleIndexPermissionStateClass {
+  indexPatterns: OptionSeletion;
+  dls: string;
+  flsMethod: 'exclude' | 'include';
+  fls: OptionSeletion;
+  maskedFields: OptionSeletion;
+  allowedActions: OptionSeletion;
+}
+
+function getEmptyIndexPermission(): RoleIndexPermissionStateClass {
+  return {
+    indexPatterns: [],
+    dls: '',
+    flsMethod: 'exclude',
+    fls: [],
+    maskedFields: [],
+    allowedActions: [],
+  };
+}
+
+function buildIndexPermissionState(
+  indexPerm: RoleIndexPermission[]
+): RoleIndexPermissionStateClass[] {
+  return indexPerm.map(perm => ({
+    indexPatterns: perm.index_patterns.map(stringToComboBoxOption),
+    dls: perm.dls,
+    // Leading ~ indicates exclude.
+    flsMethod: perm.fls.some((s: string) => !s.startsWith('~')) ? 'exclude' : 'include',
+    fls: perm.fls.map((s: string) => s.replace(/^~/, '')).map(stringToComboBoxOption),
```

Comment:
appreciate if you could naming these two arrow function for readability.

Reply:
btw

```!s.startsWith('~')) ? 'exclude' : 'include'```
assuming above is same as, right?
```s.startsWith('~')) ? 'include' : 'exclude'```

if so, this is more readable to me, just suggestion, not required to change 
``` 
function filterIncludeFoo(s) {
  return s.startsWith('~')) ? 'include' : 'exclude';
}
```

Reply:
Oh it's a bug. `~` should be `'exclude'`.

Reply:
Bug fixed.